### PR TITLE
Update applesingle.c so it builds under linux

### PR DESCRIPTION
**Pull request by MoralCode**

### CODE REVIEW

File: applesingle.c [hunk 1]
```diff
@@ -1278,7 +1278,7 @@ int main(int argc, char * argv[])
 "Encoding of files is not supported on this platform.\n"
 "\n"
 #endif
-, me, me, me, me, me);
+, me, me, me);
```
Comment (fthain) [inline applesingle.c:1281]: A compiler warning is expected here when !defined(ENABLE_ENCODING). Your patch would introduce a real bug when defined(ENABLE_ENCODING).

File: applesingle.c [hunk 2]
```diff
@@ -915,7 +915,7 @@ static int output_digest(FILE *f, char *buf, unsigned buf_sz, size_t n)
 		n -= r;
 	}
 	MD5_Final((void*)md, &c);
-	printf("%016llx%016llx (md5)\n", md[0], md[1]);
+	printf("%016lx%016lx (md5)\n", md[0], md[1]);
```
Comment (fthain) [inline applesingle.c:918]: That change produces a warning when building on 32-bit Linux/x86:
`warning: format '%lu' expects argument of type 'long unsigned int', but argument 3 has type 'UInt64' {aka 'long long unsigned int'}`
I'm not sure what the right fix is. Perhaps UInt64 should be changed to size_t? No idea what the format string would look like though.

File: applesingle.c [hunk 3]
```diff
@@ -8,7 +8,7 @@ Build cmd for Mac OS 10.4:
 Build cmd for later releases (e.g. with openssl installed under /usr/local):
   cc -m32 -Wno-deprecated-declarations -O2 -framework Carbon -lcrypto applesingle.c -o applesingle
 Build cmd for Linux/x86:
-  cc -O2 -lcrypto applesingle.c -o applesingle
+  cc applesingle.c -O2 -lcrypto -o applesingle
```
Comment (fthain) [inline applesingle.c:11]: Can you help me understand the purpose of this change? It seems to have no effect on gcc...

File: applesingle.c [hunk 4]
```diff
@@ -1278,7 +1278,7 @@ int main(int argc, char * argv[])
 "Encoding of files is not supported on this platform.\n"
 "\n"
 #endif
-, me, me, me, me, me);
+, me, me, me);
 		rc = 3;
```
Comment (fthain) [inline applesingle.c:1282]: This patch combines unrelated changes. I think it should be split up.

**Review by fthain — CHANGES_REQUESTED** [review]
Thanks for your interest in this code. I've left a few comments on your patches.

File: applesingle.c (hunk 1 above)
Comment (MoralCode) [reply applesingle.c:1281]: yeah this was mostly just a quick hacky fix because it was saying that there were too many parameters. maybe this part of the printf call can also be put inside the `#ifdef`'s so it uses a different # of parameters based on the value of defined(ENABLE_ENCODING)?

File: applesingle.c (hunk 2 above)
Comment (MoralCode) [reply applesingle.c:918]: huh, thats strange because it was giving me the opposite error before, essentially saying it expected `long long unsigned int` and was only being given `long unsigned int`. Does the type of the values contained in `md` also vary based on if `ENABLE_ENCODING` is defined?

File: applesingle.c (hunk 3 above)
Comment (MoralCode) [reply applesingle.c:11]: when building it the original way (on `Linux 5.4.0-58-generic #64-Ubuntu SMP x86_64 x86_64 x86_64 GNU/Linux`) i was getting the following linker error:

```
/usr/bin/ld: /tmp/cc95ZmtG.o: in function `main':
applesingle.c:(.text.startup+0xa5d): undefined reference to `MD5_Init'
/usr/bin/ld: applesingle.c:(.text.startup+0xb66): undefined reference to `MD5_Update'
/usr/bin/ld: applesingle.c:(.text.startup+0xbfa): undefined reference to `MD5_Final'
collect2: error: ld returned 1 exit status
```
after looking it up on https://stackoverflow.com/questions/1894013/how-to-use-openssl-with-gcc, it seemed like all the examples they used put the source file first, then the linker settings, and that seemed to fix it for me. I'm honestly still not sure why the different ordering fixed it for me

File: applesingle.c (hunk 4 above)
Comment (MoralCode) [reply applesingle.c:1282]: How would you prefer it to be split up? Pretty much the whole PR is changes to printf statements to correct compiler warnings. TBH it was pretty much all just to get it to work on my machine and contribute the changes back in case they are helpful to someone else so i wasnt particularly thinking of trying to build it with different flags or for other architectures

File: applesingle.c (hunk 1 above)
Comment (fthain) [reply applesingle.c:1281]: Okay. I've committed a change to along those lines to avoid the warning when !defined(ENABLE_ENCODING).

File: applesingle.c (hunk 2 above)
Comment (fthain) [reply applesingle.c:918]: The problem arises because you are using a 64-bit compiler but you aren't passing the -m32 flag. I've now committed a change to take care of the printf() format strings for the 64-bit case. But this may or may not be sufficient to achieve compatibility with 64-bit hosts. I haven't tested that.

File: applesingle.c (hunk 3 above)
Comment (fthain) [reply applesingle.c:11]: It would help to state which compiler you're using. You may be able to obtain information that with "cc -v". Anyway, I've committed a change to rearrange the linker option. Thanks for letting me know about this build failure.

**Comment by fthain** [comment]
I believe these issues have been resolved now.

File: applesingle.c (hunk 3 above)
Comment (MoralCode) [reply applesingle.c:11]: here's my `cc -v` output if that helps:
```
Using built-in specs.
COLLECT_GCC=cc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/9/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none:hsa
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 9.3.0-17ubuntu1~20.04' --with-bugurl=file:///usr/share/doc/gcc-9/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++,gm2 --prefix=/usr --with-gcc-major-version-only --program-suffix=-9 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --enable-default-pie --with-system-zlib --with-target-system-zlib=auto --enable-objc-gc=auto --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none=/build/gcc-9-HskZEa/gcc-9-9.3.0/debian/tmp-nvptx/usr,hsa --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 9.3.0 (Ubuntu 9.3.0-17ubuntu1~20.04) 
```
